File: PoC/src/Cake.java
package src;
import java.util.List;

public abstract class Cake {
  protected Dough dough;
  protected List<Fillings> fillings;
  protected List<Toppings> toppings;

  public Cake(Dough dough, List<Fillings> fillings, List<Toppings> toppings) {
    this.dough = dough;
    this.fillings = fillings;
    this.toppings = toppings;
  }
  public Dough getDough() {
    return dough;
  }

  public void setDough(Dough dough) {
    this.dough = dough;
  }

  public List<Fillings> getFillings() {
    return fillings;
  }

  public void setFillings(List<Fillings> fillings) {
    this.fillings = fillings;
  }

  public List<Toppings> getToppings() {
    return toppings;
  }

  public void setToppings(List<Toppings> toppings) {
    this.toppings = toppings;
  }

  public abstract int getNumberOfSlices();
  public abstract double getBasePrice();
  public abstract PriceModifiers getPriceModifiers();
}

File: Cake.ts
import { Dough, Fillings, PriceModifiers, Toppings } from "./types";

export abstract class Cake {
  constructor(
    public dough: Dough,
    public fillings: Fillings[],
    public toppings: Toppings[]
  ) {}

  abstract getNumberOfSlices(): number;
  abstract getBasePrice(): number;
  abstract getPriceModifiers(): PriceModifiers;
}

export class BentoCake extends Cake {
  getNumberOfSlices(): number {
      return 4;
  }

  getBasePrice(): number {
    return 58.00;
  }

  getPriceModifiers(): PriceModifiers {
    return {
      redVelvet: 5.0,
      pistachio: 8.0,
      topping: 8.0
    };
  }
}

export class MiniCake extends Cake {
  getNumberOfSlices(): number {
      return 6;
  }

  getBasePrice(): number {
    return 110.00;
  }

  getPriceModifiers(): PriceModifiers {
      return {
        redVelvet: 5.00,
        pistachio: 8.00,
        topping: 8.00
      }
  }
}

export class Cake15Slices extends Cake {
  getNumberOfSlices(): number {
      return 15;
  }

  getBasePrice(): number {
      return 160.00;
  }

  getPriceModifiers(): PriceModifiers {
      return {
        redVelvet: 10.00,
        pistachio: 14.00,
        topping: 15.00
      }
  }
}

export class Cake25Slices extends Cake {
  getNumberOfSlices(): number {
      return 25;
  }

  getBasePrice(): number {
      return 240.00;
  }

  getPriceModifiers(): PriceModifiers {
      return {
        redVelvet: 20.00,
        pistachio: 24.00,
        topping: 25.00
      }
  }
}